datadog-agent: support unprivileged build

maxgio92 opened the pull request:
This is a draft to try to support unprivileged build for the Datadog Agent.

maxgio92 committed:
wip(datadog-agent): remove sudo requirement for system_probe task invoke

Signed-off-by: Massimiliano Giovagnoli <massimiliano.giovagnoli@chainguard.dev>
Co-authored-by: Wojciech Kocjan <wojciech.kocjan@chainguard.dev>

maxgio92 commented on datadog-agent/tasks-system_probe.py.patch:
```diff
@@ -0,0 +1,20 @@
+
+         java_dir = os.path.join("pkg", "network", "protocols", "tls", "java")
+         ctx.run(f"{sudo} mkdir -p {EMBEDDED_SHARE_JAVA_DIR}")
+-        ctx.run(f"{sudo} install -m644 -oroot -groot {java_dir}/agent-usm.jar {EMBEDDED_SHARE_JAVA_DIR}/agent-usm.jar")
```
This must be analyzed and tested in depth IMO

maxgio92 commented:
It may be not needed with https://github.com/chainguard-dev/melange/pull/1572

smoser commented:
I successfully built as un-privileged bwrap runner with this change.  Its very similar to the change you're proposing but does the "patch" at a lower level. with this change, `invoke` thinks it is running as root so it won't bother sudo.



```diff
@@ -67,6 +67,7 @@ environment:
       - python3-dev~3.11 # strictly requires python3.11
       - systemd-dev
       - wget # Required for downloading clang-12 and kernel headers from debian
+      - util-linux-misc # unshare
   environment:
     # CGo allows Go programs to call C code
     CGO_ENABLED: "1"
@@ -143,7 +144,8 @@ pipeline:
 
       # Build once to correctly setup links/generates. The system-probe we end
       # up using will be part of the multicall below.
-      invoke -e system-probe.build \
+      unshare --user --map-root-user \
+        invoke -e system-probe.build \
         --strip-object-files \
         --no-bundle \
         --bundle-ebpf
```

maxgio92 commented:
Love it @smoser, that's elegant and should be transparent, thanks. I guess that no capability is needed as they'll be dropped in the new `invoke -e system-probe.build` process, because it completes succesfully.

maxgio92 committed:
refactor(datadog-agent): build system-probe via invoke in unshared userns

This allows the system-probe invoke task to avoid running sudo
and removes the requirement to build the package as real root.

Signed-off-by: Massimiliano Giovagnoli <massimiliano.giovagnoli@chainguard.dev>
Co-authored-by: Scott Moser <scott.moser@chainguard.dev>

maxgio92 commented:
I just applied your suggestion @smoser

smoser commented:
https://github.com/wolfi-dev/os/pull/33387 i put up. you can ack mine or i can ack yours.

maxgio92 commented:
Superseded by https://github.com/wolfi-dev/os/pull/33387
